docs/architecture/adrs/005-chat-completion-support.md

```diff
@@ -0,0 +1,74 @@
+# ADR 005: Design for chat completion support
+
+To enable running guardrails on chat completion use-cases, we want to provide a chat completion API that allows users to run various types of detectors in their chat completion flow.
+
+## Motivation
+
+ The [OpenAI chat completions](https://platform.openai.com/docs/guides/chat-completions) feature has been a popular implementation of using generative models to respond in chat conversations. With various conversation use cases such as customer service, we want to expand our API to not only support applying detectors on text in general, but specifically on chat messages.
+
+## Decisions
+
+### General
+- We will use the [OpenAI chat completions API](https://platform.openai.com/docs/api-reference/chat) as our the base API for chat completions. Parameters detailed here can be referenced.
+- Different chat completions server solutions like [vllm](https://docs.vllm.ai/en/stable/index.html) may provide [extra parameters](https://docs.vllm.ai/en/stable/serving/openai_compatible_server.html#extra-parameters-for-chat-api). For the orchestrator API to remain more agnostic of particular server implementation, the orchestrator can pass through any additional parameters to the generation server, which can do validation of parameters.
+- Guardrails-related API updates would be additive, and we will not modify any OpenAI API elements, like `messages` or `choices`.
+    - We will add a `detectors` block in the request, containing `input` and `output` blocks. This will provide flexibility and control, allowing users to provide list of detectors with parameters for both chat completions input and output separately.
+    - Here, `input` and `output` does not strictly refer to whether only the `input` to chat completions is provided to the detectors or that only the `output` to chat completions is provided to the detectors. Detectors specified in `output` can potentially take both input and output of a chat completions model.
```

Comment:
I like this idea, but it's inconsistent with what we currently have for `/api/v1/task/classification-with-text-generation`.
Should we look into combining `/api/v1/task/classification-with-text-generation` and `/api/v2/text/generation-detection` in the future?

Reply:
Yes, we will not be able to remove v1 API for quite a while.  We actually first need to create a `v2` replacement for text-generation and this chat completion would provide a solid framework for `v2/text-generation` when we get to it.

However, at this point, we need to be restricted with how things were working with `v1` endpoint.

Also, we do not plan to add such capability in `/generation` endpoint, since that will deviate its purpose.